### PR TITLE
[Driver][SYCL] Integration footer support

### PR DESCRIPTION
**Pull request by mdtoguchi**
Similar to the integration header, the device compilation can produce
an integration footer.  Have the driver take this generated footer and
add it to the host compilation.

The host source file is first preprocessed, then the integration footer is
appended to the host source file then compiled.

Contains the modifications to incorporate the integration footer under
and interim switch -fsycl-use-footer.  The plan is to enable this functionality
by default.

New tool named 'append-file' created which appends 2 files together.  This
is integrated into the toolchain to take the preprocessed host compilation
and appends the generated integration footer before doing another full
host compilation against that appended file.

### CODE REVIEW

**Comment by mdtoguchi** [comment]
@AGindinson, I wanted to get your opinion on this before I make additional adjustments.  Could you look it over and see what you think?  I'm basically creating a new 'tool' step which adds the integration footer to the preprocessed host source file.  This basically causes the preprocessing to occur instead of doing the combining of phases (e.g. full compile to object).  Since we need the preprocessing step first, intercepting the toolchain seemed reasonable as then the combine steps will fail because the steps don't look like they are expected too. 

If this method looks OK, I need to make additional adjustments to make sure that the post -append compilation steps can be combined so we don't have an additional IR and ASM step.

**Comment by AGindinson** [comment]
> Since we need the preprocessing step first, intercepting the toolchain seemed reasonable as then the combine steps will fail because the steps don't look like they are expected to.

The approach LGTM - I don't think there's a way to retain the collapsed nature of the compilation-to-object given the usage of an outside, non-Clang tool.

> I need to make additional adjustments to make sure that the post -append compilation steps can be combined so we don't have an additional IR and ASM step.

A brief look through the action-collapsing logic makes me believe this to be possible, but you'd definitely know better after actually implementing this "partial collapse". In case that doesn't prove possible, the resulting additional overhead would boil down to an extra command -> extra process creation. And I honestly don't see this as significant, given the performance gain that the recent integration header changes have brought.

**Comment by mdtoguchi** [comment]
The collapsing functionality that is in place already takes care of what we are looking for, so it was a matter of getting all of the test files in sync with the new job action to perform the append.

**Comment by mdtoguchi** [comment]
@AGindinson ,Well, we have an interesting situation here.  In theory my methodology is OK, as we are creating a JobAction to do the append after the preprocess.  The problem is that the copy/append is occurring when we put together the toolchain and not when the commands involved in the toolchain occur.  Maybe we need to create an actual tool that does the append?  You have any thoughts on approaches here?

**Comment by AGindinson** [comment]
> The problem is that the copy/append is occurring when we put together the toolchain and not when the commands involved in the toolchain occur. Maybe we need to create an actual tool that does the append? You have any thoughts on approaches here?

I've actually spent some time believing that a dedicated (trivial) tool would be performing the append (and had to double-check on the `ConstructJob` code to disprove this assumption). To my mind, a separate tool would be a little more intuitive to test/debug in the context of the existing driver infrastructure. However, I can now see that it wouldn't save us any system calls compared to "no tool, no Clang action collapsing", and the latter approach also looks viable to me.

**Comment by mdtoguchi** [comment]
New changes up, creating a new tool to be used to append the files.  This should complete general functionality but there are test modifications that are needed due to the additional compilation steps that have been integrated.  @AlexeySachkov , @romanovvlad , this should provide the integration footer work for specialization constants if this can help you with your library work.  @erichkeane , FYI.

File: llvm/tools/append-file/append-file.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    error("input file not found");
+
+  // Copy the input file to the output file
+  llvm::sys::fs::copy_file(Input, Output);
+  if (!AppendFile.empty()) {
+    // Append the to the output file.
+    std::ofstream OutFile(Output, std::ios_base::binary | std::ios_base::app |
+                                      std::ios_base::ate);
+    std::ifstream FooterFile(AppendFile, std::ios_base::binary);
+    OutFile << FooterFile.rdbuf();
```
Comment (bader) [inline llvm/tools/append-file/append-file.cpp:62]: So, @AGindinson requested a new tool for just this line, am I right?
@AGindinson, could you summarize pros and cons of a new tool vs original solution, please? I'd like to make sure we see it the same way.

Comment (mdtoguchi) [reply llvm/tools/append-file/append-file.cpp:62]: Additionally, using a 'no-tool' approach doesn't fit well within the way the driver tool chain behaves.  The tool chain is created first, then executed.  In-line code would have the append occur during tool chain creation instead of after the preprocessed source and integration footer are created.

Comment (AGindinson) [reply llvm/tools/append-file/append-file.cpp:62]: > In-line code would have the append occur during tool chain creation instead of after the preprocessed source and integration footer are created.

Ah, of course! I've somehow tricked myself into believing that this "wrong" flow would still work, while it clearly wouldn't.

Comment (bader) [reply llvm/tools/append-file/append-file.cpp:62]: @mdtoguchi, could you update the CODEOWNERS file and assign you and @AGindinson as code owners for a new tool, please?

It would be great to avoid splitting the actions into so simple tasks requiring additional tools. :-)

Comment (mdtoguchi) [reply llvm/tools/append-file/append-file.cpp:62]: CODEOWNERS updated.  thanks.

**Review by bader — DISMISSED** [review]

File: llvm/test/tools/append-file/Inputs/a.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+aaaaaaaaaaa
+aaaaaaaaaaa
+aaaaaaaaaaa
```
Comment (AGindinson) [inline llvm/test/tools/append-file/Inputs/a.txt:3]: I'd suggest minimizing the input files:
```suggestion
```

File: llvm/test/tools/append-file/append-file.test [hunk 3]
```diff
@@ -0,0 +1,6 @@
+-- Verify append-file creates the expected output file
+-- Append files
+RUN: append-file %S/Inputs/a.txt --append=%S/Inputs/b.txt --output=%t.txt
+
+-- Verify result
+RUN: diff %t.txt %S/Inputs/ab.txt
```
Comment (AGindinson) [inline llvm/test/tools/append-file/append-file.test:6]: If the input files are minimized to 1 line per file, `ab.txt` could be removed in favor of:
```suggestion
RUN: FileCheck %s --input-file %t.txt
CHECK: aaaaaaaaaaa
CHECK-NEXT: bbbbbbbbbbb
```

**Review by AGindinson — APPROVED** [review]
The code LGTM

File: llvm/test/tools/append-file/append-file.test (hunk 3 above)
Comment (AGindinson) [reply llvm/test/tools/append-file/append-file.test:6]: As a follow-up thought, simple `RUN: echo aaaaaaaa > %s.a.txt` commands would allow to discard the `Inputs` altogether. I'm personally fine with either approach.

**Review by sergey-semenov — DISMISSED** [review]

File: llvm/test/tools/append-file/append-file.test (hunk 3 above)
Comment (mdtoguchi) [reply llvm/test/tools/append-file/append-file.test]: Thanks for the suggestions!  I've updated the test accordingly.

**Review by AGindinson — APPROVED** [review]
Thanks!

**Review by sergey-semenov — APPROVED** [review]

**Review by bader — APPROVED** [review]
